segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,308 @@
+        Futures.loop(
+                () -> !closed.get(),
+                () -> Futures.delayedFuture(ITERATION_SLEEP, this.executor)
+                             .thenRunAsync(this::processEvents, this.executor)
+                             .thenRunAsync(this::reportMetrics, this.executor),
+                executor);
+    }
+
+    /**
+     * Returns a {@link CompletableFuture} that results from the execution of the following tasks for each EventProcessor:
+     * i) getting the last processed event offset, ii) read at most {@link EventProcessorConfig#getMaxItemsAtOnce()},
+     * iii) execute the handler function for the {@link EventProcessor}, and iv) truncate the internal Segment.
+     */
+    private CompletableFuture<Void> processEvents() {
+        final Timer iterationTime = new Timer();
+        List<CompletableFuture<Object>> futures = new ArrayList<>();
+        for (EventProcessorImpl ep : eventProcessorMap.values()) {
+            futures.add(readEventsForEventProcessor(ep)
+                    .thenApplyAsync(readResult -> applyProcessorHandler(ep, readResult), this.executor)
+                    .thenApplyAsync(readResult -> updateAndTruncateInternalSegment(ep, readResult, iterationTime), this.executor)
+                    .handleAsync((r, ex) -> {
+                        // If we got an exception different from NoDataAvailableException, report it as something is off.
+                        if (ex != null && !(ex instanceof NoDataAvailableException)) {
+                            log.warn("Processing iteration failed for EventProcessor {}, retrying.", ep.getName(), ex);
+                            ep.lastIterationLatency.set(-1); // Do not report latency, as there has been a failure.
+                            return null;
+                        }
+                        return r;
+                    }));
+        }
+        return Futures.allOf(futures);
+    }
+
+    /**
+     * Reports the metrics for each EventProcessor once the processing iteration is complete.
+     */
+    private void reportMetrics() {
+        for (EventProcessorImpl ep : eventProcessorMap.values()) {
+            SegmentStoreMetrics.outstandingEventProcessorBytes(ep.getName(), this.containerId, ep.outstandingBytes.get());
+            // Only report the last iteration processing latency if it has been successful.
+            if (ep.lastIterationLatency.get() >= 0) {
+                ep.metrics.batchProcessingLatency(ep.lastIterationLatency.get());
+            }
+        }
+    }
+
+    private EventsReadAndTruncationPoints applyProcessorHandler(EventProcessorImpl ep, EventsReadAndTruncationPoints readResult) {
+        try {
+            // This executes the handler function, which is code outside this class. Let's catch any kind of exception
+            // and re-throw it as a cause for CompletionException.
+            ep.getHandler().apply(readResult.getEventsRead());
```

Comment:
Doesn't this return a Future? You need to await its completion and take any necessary action. This may involve reworking this method and upstream methods.

Reply:
Correct, now the execution of the handler is part of the iteration execution chain.